### DISCUSSION TITLE
Dagger module best practices for naming, args, style, documenting

### DISCUSSION THREAD

**Discussion by jpadams** [dagger/dagger#12455] [Ideas]
What are good names for
- modules
- functions
- arguments

How should a module be commented?
- for short overall description
- for each function

Optional args and sane defaults preferred.

- module name: noun
- function name: verb
- arg names: abbrev? not abbreviated? optionals where possible
- version tags
- return types: lean to core types for user flexibility? avoid engine eval where possible?

Secrets handling.

Interfaces?

Lintable, if possible to ensure style is compliant with best practices with a linting module from Dagger.

**Comment by kpenfound** [comment]
Proposal:
- module names: noun, e.g. `Helm`
- function names: verb, e.g. `Package`
- arg names:
  - abbrev (`dir`)? or not abbreviated (`directory`)?
  - optionals where possible
- releasing: semver git tags, e.g. v1.2.3
- return types: 
  - lean to core types for user flexibility like `Container`, `Directory`, `File`
  - avoid engine eval where possible

**Comment by sagikazarmark** [comment]
> Proposal:

Agree with most of it.

> * abbrev (`dir`)? or not abbreviated (`directory`)?

My preference: pick one and call it idiomatic, so I don't have to waste brain power on deciding it. (It also allows others to shame me into doing "the right thing")


Some high level practices I'm interested in (although may not be lintable):

- Interface for initializing a module using software version, custom image reference and/or `Container` ([example](https://github.com/sagikazarmark/daggerverse/blob/493a0d97d9468d8dee23dec1f805268e940c386d/go/main.go#L19-L35))
- Interface for creating a defensive style API that doesn't allow you to end up in an inconsistent state by default, but allows customization (eg. by setting a custom `Container` if necessary) (aka. Open-Closed principle :) )
- Validation and error (message) practices (eg. when one of two parameters are required)

Other stuff (braindump):

- Language vs Dagger naming conventions (eg. in Go abbreviations are often written in all-caps which may mess with function name generation)
- When to use `Directory` vs `[]File`
- Dependency module names colliding with custom types in a module (aka. fantasy module names `archivist` vs `archiver`)
- How do you expose a `Container`/`File`/`Directory` from a higher level type (eg. `Package`)? (struct field vs. function)
